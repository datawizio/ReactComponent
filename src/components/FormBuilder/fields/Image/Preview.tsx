import React from "react";

import { DeleteOutlined, ExclamationCircleOutlined } from "@ant-design/icons";
import { Modal } from "antd";
import { useTranslation } from "react-i18next";

export interface PreviewProps {
  value: string;
  onDelete: () => void;
}

export const Preview: React.FC<PreviewProps> = ({ value, onDelete }) => {
  const { t } = useTranslation();

  const handleClick = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    e.stopPropagation();
    Modal.confirm({
      title: t("DELETE_CONFIRM_PHOTO"),
      icon: <ExclamationCircleOutlined />,
      okText: t("YES"),
      cancelText: t("CANCEL"),
      onOk() {
        onDelete();
      },
      onCancel() {}
    });
  };

  return (
    <div className="ant-upload-list ant-upload-list-picture-card">
      <div className="ant-upload-list-picture-card-container">
        <span>
          <div
            className="ant-upload-list-item ant-upload-list-item-done ant-upload-list-item-list-type-picture-card"
            onClick={handleClick}
          >
            <div className="ant-upload-list-item-info">
              <span>
                <img
                  src={value}
                  alt="Preview"
                  className="ant-upload-list-item-image"
                />
              </span>
            </div>
            <span className="ant-upload-list-item-actions">
              <DeleteOutlined />
            </span>
          </div>
        </span>
      </div>
    </div>
  );
};
