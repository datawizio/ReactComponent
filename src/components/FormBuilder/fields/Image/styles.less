.field-image-upload-container.ant-upload.ant-upload-drag {
  width: 150px;
  height: 150px;
  .ant-upload-btn {
    padding: 0;
  }
  .ant-upload-list.ant-upload-list-picture-card {
    .ant-upload-list-picture-card-container {
      float: none;
      margin: 0;
      .ant-upload-list-item {
        margin: 0;
        width: 148px;
        height: 148px;
        padding: 0;
        border: none;
        border-radius: 50%;
        overflow: hidden;
        .ant-upload-list-item-info {
          &::before {
            top: 0;
            left: 0;
          }
        }
      }
    }
  }
}
