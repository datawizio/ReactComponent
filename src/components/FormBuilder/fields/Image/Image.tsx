import React, { useContext } from "react";
import { ImageProps } from "../../types";
import { Upload, message } from "antd";
import { RcFile } from "antd/lib/upload";
import { PlusOutlined } from "@ant-design/icons";
import { Preview } from "./Preview";
import ImgCrop from "antd-img-crop";
import ConfigContext from "../../../ConfigProvider/context";

function beforeUpload(file: any) {
  const isJpgOrPng = file.type === "image/jpeg" || file.type === "image/png";
  if (!isJpgOrPng) {
    message.error("You can only upload JPG/PNG file!");
  }
  const isLt2M = file.size / 1024 / 1024 < 2;
  if (!isLt2M) {
    message.error("Image must smaller than 2MB!");
  }
  return isJpgOrPng && isLt2M;
}

export const Image: React.FC<ImageProps> = ({
  name,
  value,
  placeholder,
  onChange
}) => {
  const { translate } = useContext(ConfigContext);

  const upload = (file: RcFile) => {
    const reader = new FileReader();
    reader.readAsDataURL(file);
    reader.onload = () => {
      if (reader.result && onChange)
        onChange({ name, value: reader.result as string });
    };
    return "";
  };

  const handleDelete = () => {
    onChange && onChange({ name, value: null });
  };

  const uploadButton = value ? (
    <Preview value={value} onDelete={handleDelete} />
  ) : (
    <div>
      <PlusOutlined />
      <div className="ant-upload-text">{placeholder}</div>
    </div>
  );
  return (
    <ImgCrop
      shape="round"
      modalTitle={translate("EDIT_IMAGE")}
      modalOk={translate("SUBMIT")}
      modalCancel={translate("CANCEL")}
    >
      <Upload.Dragger
        beforeUpload={beforeUpload}
        action={upload}
        listType="picture-card"
        showUploadList={false}
        className="field-image-upload-container"
        customRequest={() => {}}
      >
        {uploadButton}
      </Upload.Dragger>
    </ImgCrop>
  );
};
