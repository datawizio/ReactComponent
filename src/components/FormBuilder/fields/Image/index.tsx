import React from "react";
import { FieldImageProps } from "../../types";

import { Form } from "antd";
import { Image } from "./Image";

import "./styles.less";

export const FieldImage: React.FC<FieldImageProps> = React.memo(
  ({ name, label, rules, placeholder, onChange }) => {
    return (
      <Form.Item name={name} label={label} rules={rules}>
        <Image onChange={onChange} name={name} placeholder={placeholder} />
      </Form.Item>
    );
  }
);
