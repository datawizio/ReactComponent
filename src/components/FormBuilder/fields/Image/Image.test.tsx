import "jsdom-global/register";
import React from "react";
import { shallow } from "enzyme";

import { Image } from "./Image";

const mockProps = {
  name: "Image name",
  placeholder: "placeholder",
  value: "value",
  onChange: jest.fn()
};

const setUp = (props?) => shallow(<Image {...props} />);

describe("FormBuilder Image component", () => {
  let component;
  beforeEach(() => {
    component = setUp(mockProps);
  });

  it("Image rendered correctly", () => {
    expect(component).toMatchSnapshot();
  });

  it("Image call upload action", () => {
    const mockImg = {
      type: "image/jpeg",
      size: 10000
    };
    const mockInvalidImg = {
      type: "invalid/type",
      size: 10000
    };

    const mockTooLargeImg = {
      type: "image/jpeg",
      size: 5000000
    };
    const uploadDragger = component
      .find(".field-image-upload-container")
      .first();
    const isValidImg = uploadDragger.props().beforeUpload(mockImg);
    const isInvalidImg = !uploadDragger.props().beforeUpload(mockInvalidImg);
    const isTooLarge = !uploadDragger.props().beforeUpload(mockTooLargeImg);

    expect(isValidImg).toBeTruthy();
    expect(isInvalidImg).toBeTruthy();
    expect(isTooLarge).toBeTruthy();
  });

  it("Image call upload action", () => {
    Object.defineProperty(global, "FileReader", {
      writable: true,
      value: jest.fn().mockImplementation(() => ({
        readAsDataURL: jest.fn(),
        onLoad: jest.fn(),
        result: true
      }))
    });

    const mockFile = {
      uid: "id"
    };
    const uploadDragger = component
      .find(".field-image-upload-container")
      .first();
    const uploadResult = uploadDragger.props().action(mockFile);
    expect(uploadResult).toBe("");
  });
});
