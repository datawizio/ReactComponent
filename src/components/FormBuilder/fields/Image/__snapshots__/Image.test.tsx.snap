// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormBuilder Image component Image rendered correctly 1`] = `
<ForwardRef
  aspect={1}
  fillColor="white"
  grid={false}
  maxZoom={3}
  minZoom={1}
  modalCancel="CANCEL"
  modalOk="SUBMIT"
  modalTitle="EDIT_IMAGE"
  quality={0.4}
  rotate={false}
  shape="round"
  zoom={true}
>
  <Dragger
    action={[Function]}
    beforeUpload={[Function]}
    className="field-image-upload-container"
    customRequest={[Function]}
    listType="picture-card"
    showUploadList={false}
  >
    <Preview
      onDelete={[Function]}
      value="value"
    />
  </Dragger>
</ForwardRef>
`;
